includes/discovery/bgp-peers.inc.php

```diff
@@ -11,15 +11,94 @@
+            if (empty($vrfId)) {
+                $checkVrf = ' AND `vrf_id` IS NULL ';
+            }
+            foreach ($vrf as $address => $value) {
+                $astext = get_astext($value['tBgpPeerNgPeerAS4Byte']);
+                if (dbFetchCell('SELECT COUNT(*) from `bgpPeers` WHERE device_id = ? AND bgpPeerIdentifier = ? ' . $checkVrf, [$device['device_id'], $address, $vrfId]) < '1') {
+                    $peers = [
+                        'device_id' => $device['device_id'],
+                        'vrf_id' => $vrfId,
+                        'bgpPeerIdentifier' => $address,
+                        'bgpPeerRemoteAs' => $value['fbBgpPeerRemoteAS'],
+                        'bgpPeerState' => 'idle',
+                        'bgpPeerAdminStatus' => 'stop',
+                        'bgpLocalAddr' => '0.0.0.0',
+                        'bgpPeerRemoteAddr' => '0.0.0.0',
+                        'bgpPeerInUpdates' => 0,
+                        'bgpPeerOutUpdates' => 0,
+                        'bgpPeerInTotalMessages' => 0,
+                        'bgpPeerOutTotalMessages' => 0,
+                        'bgpPeerFsmEstablishedTime' => 0,
+                        'bgpPeerInUpdateElapsedTime' => 0,
+                        'astext' => $astext,
+                    ];
+                    dbInsert($peers, 'bgpPeers');
+                    if (Config::get('autodiscovery.bgp')) {
+                        $name = gethostbyaddr($address);
+                        discover_new_device($name, $device, 'BGP');
+                    }
+                    echo '+';
+                } else {
+                    dbUpdate(['bgpPeerRemoteAs' => $value['fbBgpPeerRemoteAS'], 'astext' => $astext], 'bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
+                    echo '.';
+                }
+            }
+        }
+        // clean up peers
+        $peers = dbFetchRows('SELECT `B`.`vrf_id` AS `vrf_id`, `bgpPeerIdentifier`, `vrf_oid` FROM `bgpPeers` AS B LEFT JOIN `vrfs` AS V ON `B`.`vrf_id` = `V`.`vrf_id` WHERE `B`.`device_id` = ?', [$device['device_id']]);
+        foreach ($peers as $value) {
+            $vrfId = $value['vrf_id'];
+            $vrfOid = $value['vrf_oid'];
+            $address = $value['bgpPeerIdentifier'];
+
+            if (empty($bgpPeers[$vrfOid][$address])) {
+                $deleted = dbDelete('bgpPeers', 'device_id = ? AND bgpPeerIdentifier = ? AND vrf_id = ?', [$device['device_id'], $address, $vrfId]);
+
+                echo str_repeat('-', $deleted);
+                echo PHP_EOL;
+            }
+        }
+    }
+
+    if ($device['os'] != 'firebrick') {
+        unset($bgpPeers);
+    }
 
     if (key_exists('vrf_lite_cisco', $device) && (count($device['vrf_lite_cisco']) != 0)) {
         $vrfs_lite_cisco = $device['vrf_lite_cisco'];
     } else {
         $vrfs_lite_cisco = [['context_name'=>'']];
     }
 
```

Comment:
The block below can be avoided if you define bgpLocalAs in your dedicated discovery block.

Reply:
Noted - and added - although there's still some work around this to be done on the Firebrick side to be sensible.